tests/core/test_entity.py

```diff
@@ -738,3 +725,458 @@ def test_first_with_cache(self):
+        # Filter by the Owner
+        q1 = Dog.query.filter(Q(owner='John'))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (AND: ('owner', 'John'))>,)"
+        assert decon_kwargs == {}
+        assert q1._criteria.connector == Q.AND
+        assert q1._criteria.negated is False
+        assert Q(*decon_args, **decon_kwargs) == q1._criteria
+
+    def test_exclude_with_single_q_object(self):
+        """Test query construction with negation"""
+        # Filter by the Owner
+        q1 = Dog.query.exclude(Q(owner='John'))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (AND: ('owner', 'John'))>,)"
+        assert decon_kwargs == {'_negated': True}
+        assert q1._criteria.connector == Q.AND
+        assert q1._criteria.negated is True
+        assert Q(*decon_args, **decon_kwargs) == q1._criteria
+
+    def test_filter_with_multiple_q_objects(self):
+        """Test query construction with negation"""
+        # Filter by the Owner
+        q1 = Dog.query.filter(Q(owner='John'), Q(age=3))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (AND: ('owner', 'John'))>, <Q: (AND: ('age', 3))>)"
+        assert decon_kwargs == {}
+        assert q1._criteria.connector == Q.AND
+        assert q1._criteria.negated is False
+        assert Q(*decon_args, **decon_kwargs) == q1._criteria
+
+    def test_exclude_with_multiple_q_objects(self):
+        """Test query construction with negation"""
+        q1 = Dog.query.exclude(Q(owner='John'), Q(age=3))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (AND: ('owner', 'John'))>, <Q: (AND: ('age', 3))>)"
+        assert decon_kwargs == {'_negated': True}
+        assert q1._criteria.connector == Q.AND
+        assert q1._criteria.negated is True
+        assert Q(*decon_args, **decon_kwargs) == q1._criteria
+
+    def test_filter_with_and(self):
+        """Test query construction with negation"""
+        # Filter by the Owner
+        q1 = Dog.query.filter(Q(owner='John') & Q(age=3))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (AND: ('owner', 'John'), ('age', 3))>,)"
+        assert decon_kwargs == {}
+        assert q1._criteria.connector == Q.AND
+        assert q1._criteria.negated is False
+        assert Q(*decon_args, **decon_kwargs) == q1._criteria
+
+    def test_filter_with_or(self):
+        """Test query construction with negation"""
+        # Filter by the Owner
+        q1 = Dog.query.filter(Q(owner='John') | Q(age=3))
+
+        _, decon_args, decon_kwargs = q1._criteria.deconstruct()
+
+        assert str(decon_args) == "(<Q: (OR: ('owner', 'John'), ('age', 3))>,)"
+        assert decon_kwargs == {}
+        assert q1._criteria.connector == Q.AND
```

Comment:
Should this be an `AND`

Reply:
`Q(owner='John') | Q(age=3)` has an `OR` connector alright, but on passing the criteria to filtering methods, we do `clone._add_q(Q(*args, **kwargs))` and that adds a surrounding Q object. This method is efficient because we can treat Q objects, args, and kwargs in the same manner, but has the side effect of adding an `AND` connector on the outer Q object: `<Q: (AND: (OR: ('owner', 'John'), ('age', 3)))>`.

I was thinking about this, and I realized the check for the connector, if we do it at all, should happen one level deeper, always. I should probably get rid of this check in all test cases for now.

Reply:
I have changed the test cases to check for the connector at the right level now. 👍 Good catch, buddy.